pkg/vm/engine/tae/model/aot.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine/tae/containers"
+	"github.com/tidwall/btree"
+)
+
+type RowsT[T any] interface {
+	Length() int
+	Window(offset, length int) T
+}
+
+type BlockT[R RowsT[R]] interface {
+	Append(R) error
+	IsAppendable() bool
+	Length() int
+	String() string
+	Close()
+}
+
+type AOT[B BlockT[R], R RowsT[R]] struct {
+	sync.Mutex
+	blockSize int
+	appender  B
+	blocks    *btree.BTreeG[B]
+	factory   func(R) B
+}
+
+func NewAOT[B BlockT[R], R RowsT[R]](
+	blockSize int,
+	factory func(R) B,
+	lessFn func(_, _ B) bool) *AOT[B, R] {
+	return &AOT[B, R]{
+		blockSize: blockSize,
+		factory:   factory,
+		blocks:    btree.NewBTreeGOptions(lessFn, btree.Options{NoLocks: true}),
+	}
+}
+
+func (aot *AOT[B, R]) Iter() btree.GenericIter[B] {
```

Comment:
Returning a btree.GenericIter is almost surely wrong.   If you have to use an iter, write our own interface and use our own

Reply:
Not used right now and removed.